pkg/statsd/dispatcher.go

```diff
@@ -51,28 +52,30 @@ func NewMetricDispatcher(numWorkers int, perWorkerBufferSize int, af AggregatorF
 }
 
 // Run runs the MetricDispatcher.
-func (d *MetricDispatcher) Run(ctx context.Context, done gostatsd.Done) {
-	defer done()
-	var wg sync.WaitGroup
-	wg.Add(d.numWorkers)
-	for _, worker := range d.workers {
-		w := worker // Make a copy of the loop variable! https://github.com/golang/go/wiki/CommonMistakes
-		go w.work(wg.Done)
-	}
+func (d *MetricDispatcher) Run(ctx context.Context) {
+	var wg wait.Group
 	defer func() {
 		for _, worker := range d.workers {
 			close(worker.metricsQueue) // Close channel to terminate worker
 		}
 		wg.Wait() // Wait for all workers to finish
 	}()
+	for _, worker := range d.workers {
+		wg.Start(worker.work)
+	}
 
 	// Work until asked to stop
 	<-ctx.Done()
 }
 
-func (d *MetricDispatcher) runMetrics(ctx context.Context, statser statser.Statser) {
+func (d *MetricDispatcher) RunMetrics(ctx context.Context, statser statser.Statser) {
```

Comment:
Why not?

Well, that is how I think software of high quality should be written in 99% of cases - everything, every module is a library. Only exception is `main` package (also "special" are things like OS kernel, drivers, firmware, etc). Every module should provide and API surface and make 0 assumptions about how it is used.

Let me explain why I think this is the right way. Lets take a look at the opposite to see why it is bad :)

1. git is a good example of the opposite approach. Every subcommand is a program (that is fine!) and that program is written without any internal boundaries/abstrations/modules AFAIK from some blopost, cannot find it. E.g. if there is an error it just terminates the whole process. Cannot use such code as library. Found this https://githubengineering.com/move-fast/

> With these limitations in mind, five years ago we began the development of libgit2, a re-implementation of most of Git’s low level APIs as a library that could be linked and used directly within a server-side process.

So because original authors wrote it as a program, someone had to invest enormous effort to rewrite it as a library.

2. Kuberetes has a few places where they use a "registry" pattern. A global structure where e.g. implementations of something register themselves. They are fixing this because initial design makes it hard to re-use the code in new ways that they want to use it. Also third parties are having a hard time to re-use the code!

3. Gostatsd actually used to do this "registry" pattern too. I changed it to not do it. Now every backend/cloud provider can be used as a library somewhere else and it will not pull in the registry itself into the binary. See https://github.com/atlassian/gostatsd/pull/4/commits/24f4ef4abf0bedb3254a2b7151730b65aab6d1ae

Less assumptions/knowledge about global context lead to cleaner designs and smaller dependency graphs. And in the end the software is much more reusable, testable and easier to maintain and understand. And has fewer bugs.

The only downside is that sometimes this requires a bit more thinking and coding. But in the long run I've never seen it to be a bad thing or cause any issues.